### PR TITLE
feat: add table components

### PR DESCRIPTION
**Pull request by smmoosavi**

### CODE REVIEW

File: src/index.test.tsx [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    name: 'Keyboard',
+    price: 116,
+    count: 1,
+    color: 'silver',
+  },
+];
+
+describe('ctablex', () => {
+  it('should render table', () => {
+    const c = (
```
Comment (c0m1t) [inline src/index.test.tsx]: Maybe a better variable name or directly call `render` function with it

File: src/index.test.tsx [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+          <TableBody>
+            <Rows>
+              <Row />
+            </Rows>
+          </TableBody>
+        </Table>
+      </DataTable>
+    );
+    const { unmount } = render(c);
+    unmount();
```
Comment (c0m1t) [inline src/index.test.tsx]:
```suggestion
expect(screen.queryByText(/1./)).toBeInTheDocument()

File: package.json [hunk 3]
```diff
@@ -26,9 +26,17 @@
     "prepublishOnly": "yarn validate && yarn build"
   },
   "devDependencies": {
+    "@testing-library/react": "^10.3.0",
```
Comment (c0m1t) [inline package.json:48]: I think it'd be a good idea to add `@testing-library/jest-dom` as well. We could use a couple of more helpful jest matchers.

File: src/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+            <TableBody>
+              <Rows>
+                <Row />
+              </Rows>
+            </TableBody>
+          </Table>
+        </DataTable>
+      </TableUIAdapterContextProvider>
+    );
+    const { unmount } = render(c);
```
Comment (c0m1t) [inline src/index.test.tsx]:
```suggestion
expect(table).toHaveStyle({backgroundColor: '#ccc'})
```

- Also you can assign a `role` or `data-testid` to table in order to  find it easier.
- And run this test for all the adapter components

File: src/index.test.tsx [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+            <TableBody>
+              <Row />
+            </TableBody>
+          </Table>
+        </DataTable>,
+      ),
+    ).toThrow();
+    expect(() => render(<IndexCell />)).toThrow();
+    expect(() => render(<DataTable />)).toThrow();
+    render(
```
Comment (c0m1t) [inline src/index.test.tsx]:
```typescript
expect(...).not.toThrow()
```

File: src/utils/path.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+export function path<T, O>(path: string[], obj: O): T {
```
Comment (c0m1t) [inline src/utils/path.ts:1]: Can generic `O` be stricter ? i.e. to be an object?

File: src/index.ts [hunk 7]
```diff
@@ -1,12 +1,11 @@
-// read more about doc comment syntax in https://api-extractor.com/pages/tsdoc/doc_comment_syntax/
+export { DataTable } from './DataTable';
```
Comment (c0m1t) [inline src/index.ts]: Export contexts as well?

File: src/TableUIAdapterContext.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import React, {
+  createContext,
+  ElementType,
+  PropsWithChildren,
+  useContext,
+  useMemo,
+} from 'react';
+
+export interface TableUIAdapterContextType {
```
Comment (c0m1t) [inline src/TableUIAdapterContext.tsx]: Rename `TableUIAdapter` to `TableComponents`

File: src/Rows.tsx [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+interface OwnProps<D> {
+  keyAccessor?: Accessor<D, string | number>;
+  data?: ReadonlyArray<D>;
+}
+
+type Props<D> = PropsWithChildren<OwnProps<D>>;
+
+export function Rows<D>(props: Props<D>) {
+  const { children } = props;
```
Comment (c0m1t) [inline src/Rows.tsx]:
```suggestion
  const { children, keyAccessor } = props;
```
And remove line 16

File: src/index.test.tsx [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+          <Column header="Name" accessor="name">
+            <StringCell />
+          </Column>
+        </Columns>
+        <Table>
+          <TableHeader>
+            <HeaderRow />
+          </TableHeader>
+          <TableBody>
+            <Rows keyAccessor="id">
```
Comment (c0m1t) [inline src/index.test.tsx]: I think you can pass a mocked function to `keyAccessor` and see how many times and with what values it has been called

File: src/index.test.tsx [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+          </TableHeader>
+          <TableBody>
+            <Rows>
+              <Row />
+            </Rows>
+          </TableBody>
+        </Table>
+      </DataTable>
+    );
+    const { unmount } = render(c);
```
Comment (c0m1t) [inline src/index.test.tsx]: You should expect it renders one `th` and three `td`s all with no content

File: src/index.test.tsx [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    unmount();
+  });
+  it('should handle single row', () => {
```
Comment (c0m1t) [inline src/index.test.tsx]:
```suggestion
  it('should render a custom Row with external data', () => {
```
And maybe pass an object to the `<Row />`

File: src/index.test.tsx [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+  {
+    id: '3',
+    name: 'Keyboard',
+    price: 116,
+    count: 1,
+    color: 'silver',
+  },
+];
+
+describe('ctablex', () => {
```
Comment (c0m1t) [inline src/index.test.tsx:77]: - Add a test for a function `accessor` for `Column` component
- Add a test for `findColumns`
- Add a test for `ArrayOutput`
- Add a test for `getValue`

File: src/utils/accessor.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import { path } from './path';
```
Comment (c0m1t) [inline src/utils/accessor.ts]: Rename this file to `getValue`

File: src/utils/accessor.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+export type Accessor<D, C> = string | ((row: D) => C);
```
Comment (c0m1t) [inline src/utils/accessor.ts:1]: Maybe export it from a separate file? Cuase this is a type used in several files

File: index.d.ts [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * @returns the `ctablex`  string
+ *
+ * @example
+ * Here's an example:
+ *
+ * ```ts
+ * import { ctablex } from '@ctablex/core'
+ *
+ * console.log(ctablex());
```
Comment (Pymossy) [inline index.d.ts]: Let's change it to a real example!

File: src/types.d.ts [hunk 17]
```diff
@@ -0,0 +1 @@
+export { Options } from './ctablex';
```
Comment (Pymossy) [inline src/types.d.ts]: I think we should change it to `export {};`

File: src/TableComponentsContext.tsx [hunk 18]
```diff
@@ -32,13 +32,17 @@ export function useTableComponentsContext(): TableComponentsContextType {
   return useContext(TableComponentsContext);
 }
 
-interface OwnProps {
+interface TableComponentsContextProviderOwnProps {
   value: Partial<TableComponentsContextType>;
 }
 
-export type Props = PropsWithChildren<OwnProps>;
+export type TableComponentsContextProviderProps = PropsWithChildren<
```
Comment (c0m1t) [inline src/TableComponentsContext.tsx:39]: I still believe 
```js
export interface TableComponentsContextProviderProps extends PropsWithChildren<{}> {
  value: Partial<TableComponentsContextType>;
}
```

or 

```js
export type TableComponentsContextProviderProps = PropsWithChildren<{
 value: Partial<TableComponentsContextType>;
}> 
```

would be better than using `OwnProps` and `Props`

File: src/TableComponentsContext.tsx [hunk 19]
```diff
@@ -0,0 +1,56 @@
+import React, {
+  Context,
+  createContext,
+  ElementType,
+  PropsWithChildren,
+  useContext,
+  useMemo,
+} from 'react';
```
Comment (c0m1t) [inline src/TableComponentsContext.tsx:8]:
```suggestion
import * as React from 'react';
```

File: src/cell/Cell.tsx [hunk 20]
```diff
@@ -0,0 +1,18 @@
+import React, { PropsWithChildren } from 'react';
+import { useTableComponentsContext } from '../TableComponentsContext';
+import { ContentValue } from '../content/ContentValue';
+
+interface CellOwnProps<D, C> {
+  accessor: string | ((row: D) => C);
+}
+
+export type CellProps<D, C> = PropsWithChildren<CellOwnProps<D, C>>;
+export function Cell<D, C>(props: CellProps<D, C>) {
```
Comment (c0m1t) [inline src/cell/Cell.tsx]:
```suggestion
export type CellProps<D, C> = PropsWithChildren<CellOwnProps<D, C>>;

export function Cell<D, C>(props: CellProps<D, C>) {
```

File: src/cell/Cell.tsx [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+  const Components = useTableComponentsContext();
+  const { accessor, children } = props;
```
Comment (c0m1t) [inline src/cell/Cell.tsx]: More consistency in code, always destruct `props`  at the first line? 
```suggestion
  const { accessor, children } = props;
  const Components = useTableComponentsContext();
```

File: src/content/ContentValue.tsx [hunk 22]
```diff
@@ -0,0 +1,26 @@
+import React, { PropsWithChildren } from 'react';
+import { useRowData } from '../row/RowDataContext';
+import { Accessor } from '../types';
+import { getValue } from '../utils/getValue';
+import { ContentContextProvider } from './ContentContext';
+import { DefaultContent } from './DefaultContent';
+
+interface ContentValueOwnProps<D, C> {
```
Comment (c0m1t) [inline src/content/ContentValue.tsx:7]: Move it after `useContentValue` ?

File: src/utils/accessor.ts [hunk 23]
```diff
@@ -0,0 +1 @@
...
```
Comment (c0m1t) [inline src/utils/accessor.ts:1]: Add default value to `C` ?

File: src/utils/getValue.ts [hunk 24]
```diff
@@ -0,0 +1,14 @@
+import { Accessor } from './accessor';
+import { path } from './path';
+
+export function getValue<D, C>(value: D, accessor: ''): null;
```
Comment (c0m1t) [inline src/utils/getValue.ts]: Can we change `''` to something else? maybe `null`?

File: src/index.ts [hunk 25]
```diff
@@ -1,12 +1,57 @@
...
-export { ctablex } from './ctablex';
+export { DataTable } from './data/DataTable';
+export {
+  TableComponentsContext,
+  useTableComponentsContext,
+  TableComponentsContextProvider,
+} from './TableComponentsContext';
+export { useData, DataContext, DataContextProvider } from './data/DataContext';
+
+export { Columns } from './column/Columns';
```
Comment (c0m1t) [inline src/index.ts:9]: Add `index.ts` to each folder and `import * from '...' `?

File: src/data/DataTable.tsx [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+import { DataContextProvider, useData } from './DataContext';
+
+interface DataTableOwnProps<D> {
+  data?: ReadonlyArray<D>;
+}
+
+export type DataTableProps<D> = PropsWithChildren<DataTableOwnProps<D>>;
+
+export function DataTable<D>(props: DataTableProps<D>) {
+  const columns = findColumns(props.children);
```
Comment (c0m1t) [inline src/data/DataTable.tsx]:
```js
const { children } = props;
```

File: src/utils/accessor.ts (hunk 23 above)
Comment (smmoosavi) [reply src/utils/accessor.ts:1]: No

File: src/utils/getValue.ts (hunk 24 above)
Comment (smmoosavi) [reply src/utils/getValue.ts]: Maybe no. I am not sure.

File: src/TableComponentsContext.tsx (hunk 19 above)
Comment (smmoosavi) [reply src/TableComponentsContext.tsx:8]: No. it causes React_2 appeared in `index.d.ts `

```diff
+import * as React_2 from 'react';
```

```diff
-declare type CellProps<D, C> = PropsWithChildren<CellOwnProps<D, C>>;
+declare type CellProps<D, C> = React_2.PropsWithChildren<CellOwnProps<D, C>>;
```

**Review by c0m1t — APPROVED** [review]

**Pull request merged**